web/modules/custom/joinup_core/src/EventSubscriber/GroupInvitationSubscriber.php

```diff
@@ -0,0 +1,142 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\joinup_core\EventSubscriber;
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\joinup_invite\Event\InvitationEventInterface;
+use Drupal\joinup_invite\Event\InvitationEvents;
+use Drupal\og\MembershipManagerInterface;
+use Drupal\og\OgMembershipInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Event subscriber handling invitations to groups.
+ */
+class GroupInvitationSubscriber implements EventSubscriberInterface {
+
+  use StringTranslationTrait;
+
+  /**
+   * The membership manager service.
+   *
+   * @var \Drupal\og\MembershipManagerInterface
+   */
+  protected $membershipManager;
+
+  /**
+   * Constructs a new InvitationSubscriber.
+   *
+   * @param \Drupal\og\MembershipManagerInterface $membershipManager
+   *   The membership manager service.
+   */
+  public function __construct(MembershipManagerInterface $membershipManager) {
+    $this->membershipManager = $membershipManager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents(): array {
+    $events[InvitationEvents::ACCEPT_INVITATION_EVENT] = ['acceptInvitation'];
+    $events[InvitationEvents::REJECT_INVITATION_EVENT] = ['rejectInvitation'];
+
+    return $events;
+  }
+
+  /**
+   * Accepts invitations to join a group.
+   *
+   * @param \Drupal\joinup_invite\Event\InvitationEventInterface $event
+   *   The event that was fired.
+   */
+  public function acceptInvitation(InvitationEventInterface $event): void {
+    $invitation = $event->getInvitation();
+
+    // Ignore invitations to other content entities.
+    if ($invitation->bundle() !== 'group') {
+      return;
+    }
+
+    $user = $invitation->getRecipient();
+    $group = $invitation->getEntity();
+
+    if ($this->getMembership($group, $user)) {
+      drupal_set_message($this->t('There is no action pending for this user.'));
+      // If a user already has a membership and is here, it means that there was
+      // a leftover pending invitation (the subscriber is called after a pending
+      // invitation is accepted). In that case, delete the invitation. The
+      // reason behind it is that if the invitation is not deleted, it might
+      // falsely seem that the membership exists due to an accepted invitation.
+      $invitation->delete();
```

Comment:
Also, we don't delete invitations. For now we keep them (possibly in the future we'll have a purge policy in place). For this reason let's add a new invitation status, something like "not_applicable" (or a better name). and we set this status here + save.

Reply:
Not addressed.